server/server.go

```diff
@@ -742,6 +738,36 @@ type subSentAndAck struct {
 	applying bool
 }
 
+// Returns the total number of subscriptions (including offline (queue) durables).
+// This is used essentially for the monitoring endpoint /streaming/serverz.
+func (s *StanServer) numSubs() int {
+	total := 0
+	s.channels.RLock()
+	for _, c := range s.channels.channels {
+		c.ss.RLock()
+		total += len(c.ss.psubs)
+		// Need to add offline durables
+		for _, sub := range c.ss.durables {
+			if sub.ClientID == "" {
+				total++
+			}
+		}
+		for _, qsub := range c.ss.qsubs {
+			qsub.RLock()
+			total += len(qsub.subs)
+			// If this is a durable queue subscription and all members
+			// are offline, qsub.shadow will be not nil. Report this one.
+			if qsub.shadow != nil {
+				total++
+			}
+			qsub.RUnlock()
+		}
+		c.ss.RUnlock()
+	}
+	s.channels.RUnlock()
```

Comment:
With cascading three tiers of locks, do we know we do not violate order somewhere else?

Reply:
We do not. This is a well known order: the channel store is the main, then under this one we can take the channel subStore lock. Then we possibly go over queue states.
We do this on message delivery, monitor gathering of subs, etc..

Reply:
ok